From 62bd00814a44ed72b8f888fa508936b2166d46b9 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 18:59:35 -0400
Subject: [fs] btrfs: Add noautodefrag mount option

Message-id: <1407179109-13714-89-git-send-email-sandeen@redhat.com>
Patchwork-id: 87056
O-Subject: [RHEL7.1 PATCH 088/422] btrfs: Add noautodefrag mount option.
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Qu Wenruo <quwenruo@cn.fujitsu.com>

commit fc0ca9af180b91aad2fbf2fe3b16a12e1e05a760
Author: Qu Wenruo <quwenruo@cn.fujitsu.com>
Date:   Mon Jan 6 09:58:26 2014 +0800

    btrfs: Add noautodefrag mount option.

    Btrfs has autodefrag mount option but no pairing noautodefrag option,
    which makes it impossible to disable autodefrag without umount.

    Signed-off-by: Qu Wenruo <quwenruo@cn.fujitsu.com>
    Reviewed-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/Documentation/filesystems/btrfs.txt b/Documentation/filesystems/btrfs.txt
index 0a561b3..4c48138 100644
--- a/Documentation/filesystems/btrfs.txt
+++ b/Documentation/filesystems/btrfs.txt
@@ -46,10 +46,12 @@ Options with (*) are default options and will not show in the mount options.
  bytes, optionally with a K, M, or G suffix, case insensitive.
  Default is 1MB.
 
+  noautodefrag(*)
   autodefrag
- Detect small random writes into files and queue them up for the
- defrag process.  Works best for small files; Not well suited for
- large database workloads.
+ Disable/enable auto defragmentation.
+ Auto defragmentation detects small random writes into files and queue
+ them up for the defrag process.  Works best for small files;
+ Not well suited for large database workloads.
 
   check_int
   check_int_data
diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index b02d25a..44513f3 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -323,7 +323,7 @@ enum {
  Opt_no_space_cache, Opt_recovery, Opt_skip_balance,
  Opt_check_integrity, Opt_check_integrity_including_extent_data,
  Opt_check_integrity_print_mask, Opt_fatal_errors, Opt_rescan_uuid_tree,
- Opt_commit_interval, Opt_barrier,
+ Opt_commit_interval, Opt_barrier, Opt_nodefrag,
  Opt_err,
 };
 
@@ -357,6 +357,7 @@ static match_table_t tokens = {
  {Opt_enospc_debug, "enospc_debug"},
  {Opt_subvolrootid, "subvolrootid=%d"},
  {Opt_defrag, "autodefrag"},
+ {Opt_nodefrag, "noautodefrag"},
  {Opt_inode_cache, "inode_cache"},
  {Opt_no_space_cache, "nospace_cache"},
  {Opt_recovery, "recovery"},
@@ -602,6 +603,11 @@ int btrfs_parse_options(struct btrfs_root *root, char *options)
    btrfs_info(root->fs_info, "enabling auto defrag");
    btrfs_set_opt(info->mount_opt, AUTO_DEFRAG);
    break;
+  case Opt_nodefrag:
+   if (btrfs_test_opt(root, AUTO_DEFRAG))
+    btrfs_info(root->fs_info, "disabling auto defrag");
+   btrfs_clear_opt(info->mount_opt, AUTO_DEFRAG);
+   break;
   case Opt_recovery:
    btrfs_info(root->fs_info, "enabling auto recovery");
    btrfs_set_opt(info->mount_opt, RECOVERY);
-- 
1.7.1